gcloud/monitoring/metric.py

```diff
@@ -0,0 +1,175 @@
+# Copyright 2016 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Metric Descriptors for the Google Monitoring API."""
+
+# Features intentionally omitted from this first version of the client library:
+#   - Creating and deleting metric descriptors.
+
+import collections
+
+from gcloud.monitoring.label import LabelDescriptor
+
+
+class MetricDescriptor(collections.namedtuple(
+        'MetricDescriptor', ('name type labels metric_kind value_type unit'
+                             ' description display_name'))):
+    """Specification of a metric type and its schema.
+
+    Metric descriptor instances are immutable.
+
+    :type name: string
+    :param name: The "resource name" of the metric descriptor. For example:
+                 ``"projects/<project_id>/metricDescriptors/<type>"``
+
+    :type type: string
+    :param type: The metric type including a DNS name prefix. For example:
+                 ``"compute.googleapis.com/instance/cpu/utilization"``
+
+    :type labels: list of :class:`~gcloud.monitoring.label.LabelDescriptor`
+    :param labels: A sequence of label descriptors specifying the labels used
+                   to identify a specific instance of this metric.
+
+    :type metric_kind: string
+    :param metric_kind: The kind of measurement. It must be one of ``"GAUGE"``,
```

Comment:
> Choosing an enum class or pattern makes the code more opinionated.

That's OK. We're opinionated.

> Using something like enum34 adds an installation dependency.

I agree -- that's an implementation detail though, so we don't need to do it that way exactly. The idea is to avoid using magic strings.

> An ordinary string literal is less typing.

Using a variable will catch silly typos, as well as make it easy to browse which things are available in consoles like IPython. `metric_kinds.<tab><tab>`... 

And then typos (`ALING_MEAN`) will give an AttributeError... This is helpful when making API calls because the other side is such a black box. Seeing an attribute error says "the problem is on your side, double check that there's no typo?"

> For all kinds of validation in general, the server is authoritative, so we should generally let the server validate.

I agree with this, which is why there's nothing stopping people from providing a string... But the recommended way being a variable name seems like a good idea...

That is, you can say `MetricDescriptor(metric_kind='GUAGE')` just as easily as `MetricDescriptior(metric_kind=MetricKinds.GUAGE)`. The latter will give you an error from your typo. For former will send the request and get back an error of some sort.